crypto/bio/bio_lib.c

```diff
@@ -796,31 +796,32 @@ static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
         return 1;
 
 #ifndef OPENSSL_NO_SOCK
-    if (BIO_get_fd(bio, &fd) > 0)
+    if (BIO_get_fd(bio, &fd) > 0 && fd < FD_SETSIZE)
         return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
 #endif
-    if (milliseconds > 1000) {
+    if (fallback_millisec > 1000) {
```

Comment:
Strictly speaking this if (fallback_millisec > 1000) looks wrong.
Why is the due time not considered when fallback millisec is <= 1000 ?

Reply:
Here we are in the fallback case where proper waiting until max_time or input is available is not possible.
The idea is that we need to poll but not too tightly (to prevent a busy loop).
So we wait typically for the given number of milliseconds (such that the granularity of the polling can be determined by the caller) but at most for one second, which is the resolution of `max_time`.
This is also (to some extent) mentioned in the doc:
```
If sockets are not available it succeeds after waiting at most given
B<fallback_milliseconds> in order to avoid a tight busy loop at the caller.
```

Reply:
but the time out check is in the if bloc,
so it will never happen.

Reply:
how I see this you wait for a small time,
try to read, that returns EAGAIN, wait more
in a loop, that while loop has a time out.
But Why is max_time only compared against time()
when the "small time" is > 1000 ???

Reply:
And, sorry, for what is this function useful when OPENSSL_NO_SOCK is defined?

Reply:
Good point - there was a bug.
And my above description was not right regarding waiting at most one second.
I've corrected the code and extended the commenting inside `bio_wait()` and the doc of `BIO_wait()`.
Hope things are clear and correct now.

Reply:
> And, sorry, for what is this function useful when OPENSSL_NO_SOCK is defined?

As I wrote, the function is designed to be useful also in the case no sockets are available, where one needs to resort to polling. The caller can decide via `fallback_milliseconds` how tight the busy loop shall be.

Reply:
as a user I would expect to exclude socket-related code when I configure with no-sock
WTF is the use case, where you would use this function instead of a simple usleep,
when it ignores the passed in bio ?

Reply:
I've meanwhile extended the documentation and comments of both `BIO_wait()` and `BIO_connect_retry()`, which should make their inter-working and use more clear.

Note that not all BIOs are socket-based, and even blocking BIOs sometimes return early (for unclear reasons) and ask for retry. For the user all that should not matter.
For non-blocking behavior a timeout value may need to be respected.

Therefore `BIO_connect_retry()` uses `BIO_wait()` to implement this transparently, for any type of BIO.
For non-blocking and potentially even non-socket BIOs it polls every 100 milliseconds and sleeps in between, avoiding a busy loop.

Moreover, the pretty convenient `BIO_wait()` is used also in `crypto/http/http_client.c` to transparently wait for new input in between `BIO_read()` or `BIO_gets()`.

Reply:
Hmm, but the BIO_connect_retry is only for sockets.

Reply:
It originally was meant for use with sockets.
Yet there might be also BIOs not involving sockets (well, of course not those implemented by `bss_conn.c`).
And as mentioned the polling is needed for sure also for non-blocking socket-based BIOs.

Reply:
Can you give more details for when we needed polling for non-blocking socket BIOs?

Reply:
Connection attempts for non-blocking (socket-based or not) BIOs typically return rather quickly. If retry is requested this would lead to a busy loop re-doing the retry immediately over and over.
I am sure about this because before I implemented the wait/sleep mechanism my CPU load went to 100% until the request max time was reached (timeout).

Reply:
you only tried localhost ?
for connections involving possibly non-reachable network addresses
things are much different.

Reply:
I agree, as for localhost it usually does not help much waiting for connection establishment.
I definitely tried for non-local severs and remember unplugging and plugging my laptop's ethernet cable (with WiFi disable) for trying this out :smile: